### ISSUE TITLE
Coq 8.10.0 breaks tactic notations

### ISSUE THREAD

**Issue by DmxLarchey** [coq/coq#10894]
#### Description of the problem

The following `Tactic Notation` was able to capture the binder `x`
in earlier versions of Coq (8.8.* or 8.9.*) but it is now broken in 8.10.0

```
Tactic Notation "bind" hyp(x) "in" uconstr(f) "as" ident(h) :=
   set (h := fun x => f).

Fact test : nat -> nat.
Proof.
  intros n.
  bind n in (n*n) as f.
  revert n; exact f.
Defined.

Print test.  
```

In 8.10.0, it seems `x` is not substituted with `n` in the expression `fun x => f`. Hence Coq tries to typecheck `fun x => n*n` and fails. Is there an alternative way to bind with tactic notations? It is very convenient and user friendly to avoid writing `fun x y => ...` when the context already tells which are the bound variables.  

Thanks in advance,

Dominique

#### Coq Version

```
$ coqtop -v
The Coq Proof Assistant, version 8.10.0 (October 2019)
compiled on Oct 14 2019 10:57:50 with OCaml 4.07.0
```

**Comment by vbgl** [comment]
Bisection leads to #7288. Ping @herbelin

**Comment by herbelin** [comment]
Indeed, I forgot in #7288 to update the typing environment for `uconstr`s with the Ltac instances of the typing context, sorry. See #10899 (strange that the instance did not occur in our tests).

**Comment by DmxLarchey** [comment]
Maybe this is what Zimmi48 was saying but the issue is not fixed in 8.10.1 even though the `main Changes` references:

- fix Ltac regression in binding free names in uconstr;

which I suppose is this particular issue. What happened here ?

**Comment by vbgl** [comment]
Please give a more precise description of the issue you face: versions 8.9.1 and 8.10.1 behave the same on this particular example.

**Comment by Zimmi48** [comment]
> what Zimmi48 was saying

I haven't said anything. I was only silently investigating cases of automated management of milestones not working.

> which I suppose is this particular issue

Yes, there is a direct link from the changelog entry to the issue.

**Comment by DmxLarchey** [comment]
Ok something very strange happened with `opam 2`. It claimed of having installed `coq 8.10.1` and `coqide 8.10.1` and indeed `coqide` was displaying `8.10.1` but when running `coqc -v` then `8.10.0` popped up and same for `coqtop`. 

Some I am going to investigate that and I will return back here when I have a proper `coq 8.10.1` installed on my computer.

**Comment by Zimmi48** [comment]
That is super strange indeed. Many users have recently faced similar troubles when updating Coq with opam (opam claiming to have done the upgrade but not having properly installed the new version). However, this is the first time I hear that it can properly upgrade CoqIDE without having properly updated Coq. @RalfJung: Could it be an instance of the bug you observed? Do you have an issue number on the opam repo for this?

**Comment by DmxLarchey** [comment]
Well I did remove completely the concerned switch and installed it all over again. Now both `coq` and `coqide` are in sync on `8.10.1`. And now the fix works indeed. Thx a lot to @herbelin for the fix. 

Sorry for the trouble but I did not think that `opam 2` mess could occur. Only when @vbgl said `8.10.1` did fix the issue that I had the idea to check for `coqc -v`.

One thing I remember is that when I upgraded `coq` from `8.10.0` to `8.10.1`, very early on indeed, `coqide` was not available in version `8.10.1` and I had to wait until it was available.

**Comment by Zimmi48** [comment]
OK, so the installation would have failed but in a non-atomic way, letting opam think that Coq 8.10.1 was installed, and later on when checking for the dependency of CoqIDE, it thought that Coq 8.10.1 was already installed so no need to reinstall it. Given that CoqIDE 8.10.1 does actually work perfectly fine with Coq 8.10.0, it is almost impossible to notice the error until you check `coqc -v`.
